src/marked.js

```diff
@@ -107,29 +115,52 @@ export function marked(src, opt, callback) {
 
   function onError(e) {
     e.message += '\nPlease report this to https://github.com/markedjs/marked.';
+
     if (opt.silent) {
-      return '<p>An error occurred:</p><pre>'
+      const msg = '<p>An error occurred:</p><pre>'
         + escape(e.message + '', true)
         + '</pre>';
+      if (opt.async) {
+        return Promise.resolve(msg);
+      }
+      return msg;
+    }
+
+    if (opt.async) {
+      return Promise.reject(e);
     }
     throw e;
   }
 
+  if (opt.async) {
```

Comment:
I also merged the changes for #2728 in this PR because I am touching all of the same code.